TRIGGERS.md

```diff
@@ -0,0 +1,198 @@
+
+
+<h2>Atomicity of triggers</h2>
+Like all keyspace events, triggers will be executed as part of the operation which triggered them. However, since the triggers are set to be executed when the keyspace event has been triggered, Many issues can be caused by having a write call performed in the middle of operating another command. [This PR](https://github.com/redis/redis/pull/11199) was issued to address the same issue with modules performing write operations on keyspace events. We suggest following the same logic as modules and share the same notification mechanism, to operate the trigger operation during the unit operation completion.
+
+<h3>SUGGESTION</h3>
+
+Although it makes sense that all the aggregated triggering events will be processed after the "atomic" execution was completed (transaction, script or function call) this does impose some limitation on the Redis trigger since it will operate on a newer version of the keyspace than at the point of the keyspace event was triggered. 
+We could consider an option to always execute the trigger right after each Redis command  was completed, even if it was issued from 
+a nested context.
+The main challenge with this option is to have a nested script calls which is currently not allowed in Redis.
+During the development of a preliminary PoC I was able to overcome this limitation by storing the current script context (and the current LUA state in case of LUA engine) and restore them back once processing the triggers execution.
+I think that we could explore the option to have triggers execution after each executed command rather than after the complete atomic execution has been completed.
+
+
+<h2>Can trigger trigger trigger?</h2>
+
+We do not want to reach a nested call of triggers and a potential endless loops. According to this design a trigger execution as a result of some keyspace event is a terminal state.
+we could keep some recursion depth counter that will stop triggering once it reaches some depth, but that would cause some unexpected behavior for the library developers, and difficulty debugging cases where triggers where not executed.
+Another important aspect to consider is if modules based keyspace events calbacks should be triggered from trigger actions and vice versa.
+In the scope of this document I will assume that the trigger based actions **WILL NOT** cause matching module calbacks to be called and that module actions performed during it's callback operation **WILL NOT** cause triggering triggers, but each will be executed only in 1 level of nesting.
+
+So for example lets say I have a module callback **cb** set on some keyspace event **e1** and a trigger action **t** registered on some event **e2**.
+
+when the event **e1** is issued the **cb** will be executed and will cause event **e2** which will **NOT** trigger **t**.
+
+Also in case **e2** is issued the **t** is being executed and cause event **e1** which will **NOT** trigger **cb**.
```

Comment:
When a module depends on a callback, this will break the module.

Reply:
It seems to me that we need to choose between a super-safe approach, which will greatly limit what triggers can do, and a super-dangerous approach, which allows triggers to do anything and demands great caution when using them.  This feels like the same choice we make between a language like Java and one like C.  (Or between using a butter knife and a chain saw.  Most of the time I want my knives to be as sharp as possible.)  It strikes me as very confusing if we pick something in the middle.

What do others want?  A bike with or without training wheels?  Safety or power?  Many of our later decisions will flow from this basic principle.